targets/TARGET_STM/can_api.c

```diff
@@ -998,6 +1003,10 @@ int can_read(can_t *obj, CAN_Message *msg, int handle)
         can->RF1R |= CAN_RF1R_RFOM1;
     }
 
+    if(rx_irq_status == ENABLED) {
+        __HAL_CAN_ENABLE_IT(&obj->CanHandle, CAN_IT_FMP0);
```

Comment:
Do we believe this is race clear? If new data has already arrived before this line, will this generate an interrupt?

Hopefully this is solid, but it's possible you may have to do

    enter critical section
    enable FMP0 interrupt
    release FIFO
    disable critical section

to make sure the thing is ready to generate interrupts as the FIFO is released. I don't know if it's worth doing that just in case.